src/activation.rs

```diff
@@ -0,0 +1,155 @@
+use wayland_client::{
+    globals::{BindError, GlobalList},
+    protocol::{wl_seat, wl_surface},
+    Dispatch, Proxy, QueueHandle,
+};
+use wayland_protocols::xdg::activation::v1::client::{xdg_activation_token_v1, xdg_activation_v1};
+
+use crate::{
+    error::GlobalError,
+    globals::{GlobalData, ProvidesBoundGlobal},
+};
+
+/// Data attached to a token request
+pub trait RequestData: Send + Sync {
+    /// App_id of the application requesting the token, if applicable
+    fn app_id(&self) -> Option<String>;
+    /// Seat and serial of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests without
+    /// recent serials. There is no way to detect this from the client-side.
+    fn serial(&self) -> Option<(&wl_seat::WlSeat, u32)>;
+    /// Surface of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests from
+    /// unfocused surfaces. There is no way to detect this from the client-side.
+    fn surface(&self) -> Option<&wl_surface::WlSurface>;
+}
+
+impl RequestData
+    for (Option<(wl_seat::WlSeat, u32)>, Option<wl_surface::WlSurface>, Option<String>)
+{
```

Comment:
Why do we even have this trait impl inside the `SCTK` sounds just wrong. Users should wrap the data if they want to, since they also will pass user data along the request. Winit does that for example to throttle requests.

Reply:
bump.

Reply:
I can move this into the second commit. It felt to me like some simpler clients might not need any additional identifying data, when requesting a token, and would profit from this implementation.

Reply:
As alternative, SCTK has this weird design where you ask without user data and with `with_user_data` functions and have a default user data like

```
struct XdgActiovTokenData {
   serial,
   seat,
   app_id,
   surface
}
```

And the trait like  `XdgActiovationTokenDataExt` so you can get it from the `U: XdgActivationTokenDataExt`.

I'm not quite sure what I like more, but the thing is that the current way is to do like I suggested. You can look at other APIs for that matter.

Reply:
See https://docs.rs/smithay-client-toolkit/latest/smithay_client_toolkit/seat/keyboard/trait.KeyboardDataExt.html

Reply:
Thanks for the hint, I wasn't aware of this. I'll look into that.

Reply:
done